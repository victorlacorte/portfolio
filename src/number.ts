type Separator = '.' | ',';

namespace Utils {
  // mostly copied from https://github.com/adamwdraper/Numeral-js/blob/master/src/numeral.js
  /**
   * Computes the multiplier necessary to make x >= 1, effectively eliminating
   * miscalculations caused by finite precision.
   */
  export function multiplier(x: number) {
    const parts = x.toString().split('.');

    return parts.length < 2 ? 1 : Math.pow(10, parts[1].length);
  }

  /**
   * Given a variable number of arguments, returns the maximum multiplier that
   * must be used to normalize an operation involving all of them.
   */
  export function correctionFactor(...args: number[]) {
    return args.reduce((acc, curr) => {
      const currMultiplier = multiplier(curr);

      return acc > currMultiplier ? acc : currMultiplier;
    }, 1);
  }

  export function add(...args: number[]) {
    const corrFactor = correctionFactor(...args);

    return (
      args.reduce((acc, curr) => acc + Math.round(corrFactor * curr), 0) /
      corrFactor
    );
  }

  export function sub(...args: number[]) {
    const [first, ...rest] = args;

    return add(first, ...rest.map((num) => -num));
  }

  export function mul(...args: number[]) {
    return args.reduce((acc, curr) => {
      const corrFactor = correctionFactor(acc, curr);

      return (
        (Math.round(corrFactor * acc) * Math.round(corrFactor * curr)) /
        Math.round(corrFactor ** 2)
      );
    }, 1);
  }

  export function div(...args: number[]) {
    return args.reduce((acc, curr) => {
      const corrFactor = correctionFactor(acc, curr);

      return Math.round(corrFactor * acc) / Math.round(corrFactor * curr);
    });
  }

  export function toFixed(x: number, precision: number) {
    const k = Math.pow(10, precision);

    return Math.round(x * k) / k;
  }

  // TODO maybe we don't need the format and toCurrency functions since the
  // spreadsheet already has them
  export function format(
    x: number,
    decimals: number,
    decimalSeparator?: Separator,
    thousandsSeparator?: Separator,
  ) {
    const num = !isFinite(x) ? 0 : x,
      prec = !isFinite(decimals) ? 0 : Math.abs(decimals),
      dec = typeof decimalSeparator === 'undefined' ? '.' : decimalSeparator,
      sep =
        typeof thousandsSeparator === 'undefined' ? ',' : thousandsSeparator,
      numAsStr = (prec ? toFixed(num, prec) : Math.round(num))
        .toString()
        .split('.');

    if (numAsStr[0].length > 3) {
      numAsStr[0] = numAsStr[0].replace(/\B(?=(?:\d{3})+(?!\d))/g, sep);
    }

    if ((numAsStr[1] || '').length < prec) {
      numAsStr[1] = numAsStr[1] || '';
      numAsStr[1] += new Array(prec - numAsStr[1].length + 1).join('0');
    }

    return numAsStr.join(dec);
  }

  export function toCurrency(x: number) {
    return format(x, 2, '.', ',');
  }
}
