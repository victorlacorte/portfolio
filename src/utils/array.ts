namespace ArrayUtils {
  export function indicesOfSorted<T>(
    arr: T[],
    compareFn = (arr: T[]) => (a: number, b: number) =>
      Number(arr[a]) - Number(arr[b]),
  ): number[] {
    const indices = Array.from({ length: arr.length }, (v, idx) => idx);

    indices.sort(compareFn(arr));
    return indices;
  }

  export function fromIndices<T>(arr: T[], indices: number[]): T[] {
    return indices.map((idx) => arr[idx]);
  }

  export function sameLength(...arrays: any[]): boolean {
    const length = arrays[0].length;

    return arrays.every((arr) => arr.length === length);
  }
}
