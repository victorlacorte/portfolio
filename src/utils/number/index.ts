export { add, div, mul, padStart, sub, toUSD, trunc } from './number';
