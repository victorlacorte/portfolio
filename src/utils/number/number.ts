import * as R from 'ramda';

type DecimalAdjust = (x: number, digits?: number) => number;

// mostly adapted from https://github.com/adamwdraper/Numeral-js/blob/master/src/numeral.js

// https://stackoverflow.com/a/48764436

function isRound(x: number, digits: number): boolean {
  const p = Math.pow(10, digits);

  return Math.round(x * p) / p === x;
}

function _decimalAdjust(type: keyof Math, x: number, digits = 0): number {
  if (isRound(x, digits)) return x;

  const p = Math.pow(10, digits);
  const m = x * p * (1 + Number.EPSILON);

  return Math[type as string](m) / p;
}

const decimalAdjust = R.curry(_decimalAdjust);

const round: DecimalAdjust = decimalAdjust('round');

export const trunc: DecimalAdjust = decimalAdjust('trunc');

/**
 * Computes the multiplier necessary to make x >= 1, effectively eliminating
 * miscalculations caused by finite precision.
 */
export function multiplier(x: number): number {
  const parts = x.toString().split('.');

  return parts.length < 2 ? 1 : Math.pow(10, parts[1].length);
}

/**
 * Given a variable number of arguments, returns the maximum multiplier that
 * must be used to normalize an operation involving all of them.
 */
export function correctionFactor(...arguments_: number[]): number {
  return arguments_.reduce((accumulator, current) => {
    const currentMultiplier = multiplier(current);

    return accumulator > currentMultiplier ? accumulator : currentMultiplier;
  }, 1);
}

export function add(...arguments_: number[]): number {
  const corrFactor = correctionFactor(...arguments_);

  return (
    arguments_.reduce(
      (accumulator, current) => accumulator + round(corrFactor * current),
      0,
    ) / corrFactor
  );
}

export function sub(...arguments_: number[]): number {
  const [first, ...rest] = arguments_;

  return add(first, ...rest.map((number) => -number));
}

export function mul(...arguments_: number[]): number {
  return arguments_.reduce((accumulator, current) => {
    const corrFactor = correctionFactor(accumulator, current);

    return (
      (round(corrFactor * accumulator) * round(corrFactor * current)) /
      round(corrFactor ** 2)
    );
  }, 1);
}

export function div(...arguments_: number[]): number {
  return arguments_.reduce((accumulator, current) => {
    const corrFactor = correctionFactor(accumulator, current);

    return round(corrFactor * accumulator) / round(corrFactor * current);
  });
}

/**
 * Possibly unnecessary:
 * numObj.toFixed([digits])
 *
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed#Description
 */
// export function toFixed(x: number, precision: number): number {
//   const k = Math.pow(10, precision);

//   return Math.round(x * k) / k;
// }

/**
 * Possibly redundant: numObj.toLocaleString('en-US', { style: 'currency', currency: 'USD' }).slice(1)
 */
// export function format(
//   x: number,
//   decimals = 0,
//   decimalSeparator = '.',
//   thousandsSeparator = ',',
// ): string {
//   if (!isFinite(x)) {
//     throw new Error(expectedFinite(x));
//   }

//   if (!isFinite(decimals)) {
//     throw new Error(expectedFinite(decimals));
//   }

//   const prec = Math.abs(decimals);
//   const numAsStr = toFixed(x, prec).toString().split('.');

//   if (numAsStr[0].length > 3) {
//     numAsStr[0] = numAsStr[0].replace(
//       /\B(?=(?:\d{3})+(?!\d))/g,
//       thousandsSeparator,
//     );
//   }

//   if ((numAsStr[1] || '').length < prec) {
//     numAsStr[1] = numAsStr[1] || '';
//     numAsStr[1] += new Array(prec - numAsStr[1].length + 1).join('0');
//   }

//   return numAsStr.join(decimalSeparator);
// }

/**
 * Returns `x` formatted in en-US USD notation.
 *
 * @param x a numeric expression to be formatted as a currency amount
 */
export function toUSD(x: number): string {
  const formatted = Math.abs(x)
    .toLocaleString('en-US', { style: 'currency', currency: 'USD' })
    .slice(1);

  return x >= 0 ? formatted : `-${formatted}`;
}

export function padStart(
  x: number,
  maxLength: number,
  fillString = '0',
): string {
  return String(x).padStart(maxLength, fillString);
}

/**
 * Returns `x` truncated to a maximum of `digits` decimal numbers.
 *
 * @param x A numeric expression
 * @param digits The maximum number of decimal digits
 */
// export function trunc(x: number, digits = 0): number {
//   const power = Math.pow(10, digits);

//   return Math.trunc(x * power) / power;
// }
