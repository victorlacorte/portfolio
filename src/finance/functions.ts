import * as R from 'ramda';
import { add, div, mul, sub, trunc } from '../utils/number';

type Base = (price: number, quantity: number, tax: number) => number;
type Profit = (
  ...params: [...base: Parameters<Base>, prevPrice: number]
) => number;

const truncate = (x: number): number => trunc(x, 4);

function _averagePrice(price: number, quantity: number, tax: number): number;
function _averagePrice(
  price: number,
  quantity: number,
  tax: number,
  prevPrice: number,
  prevQuantity: number,
): number;
function _averagePrice(
  price: number,
  quantity: number,
  tax: number,
  prevPrice?: number,
  prevQuantity?: number,
): number {
  const total = _buyTotal(price, quantity, tax);

  if (prevPrice === undefined && prevQuantity === undefined) {
    return div(total, quantity);
  }

  return div(
    add(mul(prevPrice, prevQuantity), total),
    add(prevQuantity, quantity),
  );
}

const _buyTotal: Base = (price, quantity, tax): number =>
  add(mul(price, quantity), tax);

const _sellTotal: Base = (price, quantity, tax): number =>
  sub(mul(price, quantity), tax);

const _profit: Profit = (price, quantity, tax, prevPrice) =>
  _sellTotal(sub(price, prevPrice), quantity, tax);

const _profitPercent: Profit = (price, quantity, tax, prevPrice) =>
  sub(div(_sellTotal(price, quantity, tax), mul(prevPrice, quantity)), 1);

export const averagePrice: typeof _averagePrice = R.pipe(
  _averagePrice,
  truncate,
);

export const buyTotal: typeof _buyTotal = R.pipe(_buyTotal, truncate);

export const sellTotal: typeof _sellTotal = R.pipe(_sellTotal, truncate);

export const profit: typeof _profit = R.pipe(_profit, truncate);

export const profitPercent: typeof _profitPercent = R.pipe(
  _profitPercent,
  truncate,
);

// TODO these functions are simple date utilities
// type CompareFn = (a: PositionEntry, b: PositionEntry) => number;
// const compareAscending: CompareFn = (a, b) =>
//   a.date.equals(b.date)
//     ? -1
//     : Number(a.date.toJSDate()) - Number(b.date.toJSDate());
// const compareDescending: CompareFn = (a, b) =>
//   b.date.equals(a.date)
//     ? 1
//     : Number(b.date.toJSDate()) - Number(a.date.toJSDate());
