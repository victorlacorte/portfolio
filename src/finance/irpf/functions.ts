import { div, mul } from 'src/utils/number';
import type { Position, PositionEntry } from 'src/types';

import { buyMessage, sellMessage } from './messages';

// sort descending to remove subsequent entries
// year and year - 1
export function relevantEntries(
  year: number,
  entries: PositionEntry[],
): PositionEntry[] {
  const filtered = entries
    .filter((e) => e.date.year <= year)
    .sort((a, b) => Number(b.date.toJSDate()) - Number(a.date.toJSDate()));

  const relevant: PositionEntry[] = [];

  // for (let i = 0; i < relevant.length; i++) {
  //   const curr = relevant[i];

  //   if (curr.quantity == 0 && curr.date.year <= year - 1) {
  //     return ret;
  //   }

  //   ret.push(curr);
  // }
  for (const e of filtered) {
    if (e.quantity == 0 && e.date.year <= year - 1) {
      return relevant.filter((e) => e.date.year == year);
    }

    relevant.push(e);
  }

  return relevant;
}

// const mergeEntries = (entries: PositionEntry[]) => {
//   const sorted = entries.sort(
//     (a, b) => Number(a.date.toJSDate) - Number(b.date.toJSDate()),
//   );
//   const ret = [];

//   for (let i = 0; i < sorted.length - 1; i++) {
//     if (
//       sorted[i].date.equals(sorted[i + 1].date) &&
//       ((sorted[i].buyQuantity && sorted[i + 1].buyQuantity) ||
//         (sorted[i].sellQuantity && sorted[i + 1].sellQuantity))
//     ) {
//       if (sorted[i].buyQuantity) {
//         // const [popped0] = sorted.splice(i, 1);
//         // const [popped1] = sorted.splice(i, 1); // splice mutates the array so i calc also changes

//       } else {
//       }
//       const popped = sorted;
//       i++;
//     }
//   }
// };

export function formatEntries(
  entries: PositionEntry[],
  separator = '\n',
): string {
  return entries
    .map((e) =>
      e.buyQuantity
        ? buyMessage({
            date: e.date,
            price: div(e.buyTotal, e.buyQuantity),
            quantity: e.buyQuantity,
          })
        : sellMessage({
            date: e.date,
            price: div(e.sellTotal, e.sellQuantity),
            profit: e.sellProfit,
            quantity: e.sellQuantity,
          }),
    )
    .join(separator);
}

/**
 * Sort entries descending and return the first one that satisfies
 * entry.date.year <= year or `undefined` if none are found
 *
 * @param year
 * @param entries
 */
export function lastEntryFromYear(
  year: number,
  entries: PositionEntry[],
): PositionEntry | undefined {
  const sorted = entries.sort(
    (a, b) => Number(b.date.toJSDate()) - Number(a.date.toJSDate()),
  );

  for (const e of sorted) {
    if (e.date.year <= year) {
      return e;
    }
  }
}
// Obtain:
// Value in Y - 1
// Value in Y
// Amount owned in Y
export function irpf(year: number, entries: PositionEntry[]) {
  const relevant = relevantEntries(year, entries);
  const lastFromYear = lastEntryFromYear(year, relevant);
  const lastFromPrevYear = lastEntryFromYear(year - 1, relevant);

  return {
    [year]: lastFromYear
      ? `${lastFromYear.quantity}, ${mul(
          lastFromYear.price,
          lastFromYear.quantity,
        )}`
      : 0,
    [year - 1]: lastFromPrevYear
      ? mul(lastFromPrevYear.price, lastFromPrevYear.quantity)
      : 0,
    log: formatEntries(
      relevant
        // .filter((e) => e.date.year == year)
        .sort((a, b) => Number(a.date.toJSDate()) - Number(b.date.toJSDate())),
    ),
  };
}
