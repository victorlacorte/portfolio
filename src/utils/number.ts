import { expectedFinite } from 'src/utils/messages';

// mostly copied from https://github.com/adamwdraper/Numeral-js/blob/master/src/numeral.js

/**
 * Computes the multiplier necessary to make x >= 1, effectively eliminating
 * miscalculations caused by finite precision.
 */
export function multiplier(x: number): number {
  const parts = x.toString().split('.');

  return parts.length < 2 ? 1 : Math.pow(10, parts[1].length);
}

/**
 * Given a variable number of arguments, returns the maximum multiplier that
 * must be used to normalize an operation involving all of them.
 */
export function correctionFactor(...args: number[]): number {
  return args.reduce((acc, curr) => {
    const currMultiplier = multiplier(curr);

    return acc > currMultiplier ? acc : currMultiplier;
  }, 1);
}

export function add(...args: number[]): number {
  const corrFactor = correctionFactor(...args);

  return (
    args.reduce((acc, curr) => acc + Math.round(corrFactor * curr), 0) /
    corrFactor
  );
}

export function sub(...args: number[]): number {
  const [first, ...rest] = args;

  return add(first, ...rest.map((num) => -num));
}

export function mul(...args: number[]): number {
  return args.reduce((acc, curr) => {
    const corrFactor = correctionFactor(acc, curr);

    return (
      (Math.round(corrFactor * acc) * Math.round(corrFactor * curr)) /
      Math.round(corrFactor ** 2)
    );
  }, 1);
}

export function div(...args: number[]): number {
  return args.reduce((acc, curr) => {
    const corrFactor = correctionFactor(acc, curr);

    return Math.round(corrFactor * acc) / Math.round(corrFactor * curr);
  });
}

/**
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed#Description
 */
export function toFixed(x: number, precision: number): number {
  const k = Math.pow(10, precision);

  return Math.round(x * k) / k;
}

export function format(
  x: number,
  decimals = 0,
  decimalSeparator = '.',
  thousandsSeparator = ',',
): string {
  if (!isFinite(x)) {
    throw new Error(expectedFinite(x));
  }

  if (!isFinite(decimals)) {
    throw new Error(expectedFinite(decimals));
  }

  const prec = Math.abs(decimals);
  const numAsStr = toFixed(x, prec).toString().split('.');

  if (numAsStr[0].length > 3) {
    numAsStr[0] = numAsStr[0].replace(
      /\B(?=(?:\d{3})+(?!\d))/g,
      thousandsSeparator,
    );
  }

  if ((numAsStr[1] || '').length < prec) {
    numAsStr[1] = numAsStr[1] || '';
    numAsStr[1] += new Array(prec - numAsStr[1].length + 1).join('0');
  }

  return numAsStr.join(decimalSeparator);
}

export function toCurrency(x: number): string {
  return format(x, 2, '.', ',');
}

export function padStart(x: number, places: number, fillString = '0'): string {
  return String(x).padStart(places, fillString);
}
