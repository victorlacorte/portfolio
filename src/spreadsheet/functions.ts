import { statsFrom } from 'src/finance';
import type { BuySellEvent, Transaction } from 'src/types';

import Logger from '../utils/logging';
import { mul, sub } from '../utils/number';
import { expectedPositive } from '../utils/messages';
import { profitMessage, snapshotMessage } from './logging';

type ProfitReturn = {
  total: number;
  profit: number;
  taxDeduction: number;
  log: string;
};

// `total` and `profit` return properties should be changed
export function profit(
  year: number,
  month: number,
  transactions: Transaction[],
): ProfitReturn {
  let total = 0;
  let profit = 0;
  let taxDeduction = 0;
  const logger = new Logger();

  // `taxDeduction` is not part of the profit calculation
  const handleSell: BuySellEvent = (t, stats) => {
    if (t.taxDeduction < 0) {
      throw new Error(expectedPositive(t.taxDeduction));
    }

    if (t.date.year === year && t.date.month === month) {
      const purchasedValue = mul(stats[t.ticker].averagePrice, t.quantity);
      const currProfit = sub(t.total, purchasedValue);

      total += t.total;
      profit += currProfit;
      taxDeduction += t.taxDeduction;

      logger.add(
        profitMessage({
          ...t,
          purchasedValue,
          profit: currProfit,
        }),
      );
    }
  };

  statsFrom({
    transactions,
    onSell: handleSell,
  });

  return { total, profit, taxDeduction, log: logger.join() };
}

export function snapshot(transactions: Transaction[]) {
  const control: { [ticker: string]: Logger } = {};

  const handleOperation: BuySellEvent = (t, stats) => {
    if (!Object.prototype.hasOwnProperty.call(control, t.ticker)) {
      control[t.ticker] = new Logger();
    }

    control[t.ticker].add(snapshotMessage({ ...t }));
  };

  const stats = statsFrom({
    transactions,
    onBuy: handleOperation,
    onSell: handleOperation,
  });

  return Object.keys(control).map((ticker) => {
    if (Object.prototype.hasOwnProperty.call(stats, ticker)) {
      return {
        ticker,
        purchasedQuantity: stats[ticker].purchased.quantity,
        purchasedTotal: stats[ticker].purchased.total,
        soldQuantity: stats[ticker].sold.quantity,
        soldTotal: stats[ticker].sold.total,
        averagePrice: stats[ticker].averagePrice,
        log: control[ticker].join(),
      };
    }
  });
}
