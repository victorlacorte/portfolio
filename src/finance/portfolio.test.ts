import faker from 'faker';

import SimpleDate from '../utils/date';
import { add } from '../utils/number';
import type { BuyTransaction, Position, Transaction } from '../types';

import portfolioReducer, { lastEntryFrom } from './portfolio';
import { averagePrice } from './functions';

// TODO tests did not break after modifying the portfolio sell entry

// The date and ticker are irrelevant to all tests
const date = SimpleDate.make(2020, 1, 1);
const ticker = faker.random.alphaNumeric(5);

// buy.min >= abs(sell.max) or we could run into invalid number combinations
// e.g. sell quantity > buy one
const makeBuyQuantity = (): number => faker.datatype.number({ min: 100 });
const makeSellQuantity = (): number =>
  faker.datatype.number({ min: -100, max: -1 });
const makePrice = (): number => faker.datatype.float({ min: 0 });
const makeTax = (): number => faker.datatype.float({ min: 0 });
const makeIrrf = (): number => faker.datatype.float({ min: 0 });

describe('finance/portfolioReducer', () => {
  it('Correctly reproduces the example in https://www.bussoladoinvestidor.com.br/calculo-do-preco-medio-de-acoes/', () => {
    const ticker = 'petr4';
    const t: Transaction[] = [
      {
        kind: 'buy',
        date: SimpleDate.make(2020, 2, 3),
        ticker,
        price: 9,
        quantity: 100,
        tax: 21.5,
      },
      {
        kind: 'buy',
        date: SimpleDate.make(2020, 2, 5),
        ticker,
        price: 12,
        quantity: 100,
        tax: 21.7,
      },
      {
        kind: 'sell',
        date: SimpleDate.make(2020, 2, 5),
        ticker,
        price: 11.5,
        quantity: 50,
        tax: 20.86,
        irrf: 0.01,
      },
      {
        kind: 'buy',
        date: SimpleDate.make(2020, 2, 10),
        ticker,
        price: 15,
        quantity: 200,
        tax: 23.6,
      },
    ];

    const res = t.reduce(portfolioReducer, {});

    expect(res).toMatchSnapshot();
    expect(lastEntryFrom(res, ticker).price).toBe(13.2314);
  });

  test('Throws when attempting to sell a non-existent ticker', () => {});
  test('Throws when attempting to sell an invalid amount', () => {});
});
