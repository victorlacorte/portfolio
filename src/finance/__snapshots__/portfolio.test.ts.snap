// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finance/portfolioReducer Correctly reproduces the example in https://www.bussoladoinvestidor.com.br/calculo-do-preco-medio-de-acoes/ 1`] = `
Object {
  "petr4": Array [
    Object {
      "buyQuantity": 100,
      "buyTotal": 921.5,
      "date": "2020/02/03",
      "kind": "buy",
      "price": 9.215,
      "quantity": 100,
    },
    Object {
      "buyQuantity": 100,
      "buyTotal": 1221.7,
      "date": "2020/02/05",
      "kind": "buy",
      "price": 10.716,
      "quantity": 200,
    },
    Object {
      "date": "2020/02/05",
      "kind": "sell",
      "price": 10.716,
      "quantity": 150,
      "sellIrrf": 0.01,
      "sellProfit": 18.34,
      "sellProfitPercent": 0.0342,
      "sellQuantity": 50,
      "sellTotal": 554.14,
    },
    Object {
      "buyQuantity": 200,
      "buyTotal": 3023.6,
      "date": "2020/02/10",
      "kind": "buy",
      "price": 13.2314,
      "quantity": 350,
    },
  ],
}
`;
